### PR TITLE
Use SourceSets alone to configure input .java files

### LINKED ISSUE #52: Use idiomatic sourceSets for all listings of source files
The idiomatic way in Gradle to specify what sources a project is using is to use source sets.  Usually 'main' for production code and 'test' for test code.

Currently, source files are found by aggregating globs of .java files, ex:
```
srcFiles = files(
                        fileTree(dir: projectDir,
                                include: "**/*.java",
                                exclude: project.relativePath(buildDir)) +
                        fileTree(dir: "build/source/apt",
                                include: "**/*.java")
                        )
```

Instead we should always be using the sourceSets in the project.  Currently, we will erroneously find files that are not part of the project build but just happen to be in the directory, miss files the user has added to the source set (specifically generated source directories, which have to be added manually but should not have to be), and causes special casing like the above for apt (this should also allow us to fix #45.).

In addition for test sources:
```
 // Doesn't use 'buildDir' as missing full path with --no-package-directories flag
                srcFiles = files(fileTree(dir: projectDir, includes: ["**/*Test.java"]))
```

we should be using the 'test' source set.

### LINKED ISSUE #57: Improving include and exclude patterns
Per discussion from @brunobowden  and @advayDev1, consider changing our include and exclusion patterns to something resembling:
http://gradle.org/docs/current/javadoc/org/gradle/api/tasks/util/PatternFilterable.html

See discussion at #54

### LINKED ISSUE #95: sourceTask incremental compile
This should also be helpful for include / exclude of certain source files, or is that through source sets?

### PR DESCRIPTION
**Pull request by advayDev1**
Use SourceSets alone to configure input .java files
- Translation takes main and test java sourceSets
- Translation can be filtered using j2objcConfig.translatePattern
- Test takes test java sourceSet
- Test execution can be filtered using j2objcConfig.testPattern
- Cycle finder is analogous to translation.
- Also makes translation dependent on all possible input files.
- Prohibits incremental translation with --build-closure.

Fixes #57; Fixes #52; Fixes #95
Improves #85

### CODE REVIEW

**Comment by advayDev1** [comment]
alright, after this PR goes in, I think I'm done with basically important stuff I'm responsible for at the milestone.  grouped various fixes here for my ease.

**Comment by advayDev1** [comment]
Also @confile  and @brunobowden please test this thoroughly esp. the cycleFinder and xcode tasks (I don't use them at all).  translation, compile, and test all seem to be pretty solid, with incremental builds, up-to-date checks, etc. etc.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy [hunk 1]
```diff
@@ -71,20 +72,23 @@ class J2objcPlugin implements Plugin<Project> {
...
+            }
+
             extensions.create("j2objcConfig", J2objcPluginExtension)
 
             // Produces a modest amount of output
             logging.captureStandardOutput LogLevel.INFO
 
-            tasks.create(name: "j2objcCycleFinder", type: J2objcCycleFinderTask) {
+            tasks.create(name: "j2objcCycleFinder", type: J2objcCycleFinderTask,
+                    dependsOn: 'test') {
```
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: Should this depend on `'check'`, similar to `lateDependsOn(project, 'j2objcTest', 'check')`. If you change it that way, add a comment that this is an implicit dependency on `'test'`.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy [hunk 2]
```diff
@@ -71,20 +72,23 @@ class J2objcPlugin implements Plugin<Project> {
...
         // This avoids a lot of "project." prefixes, such as "project.tasks.create"
         project.with {
-            // TODO: dependency on project.j2objcConfig, so any setting change
-            // TODO: invalidates all (ideally some) tasks and causes a rebuild
+            if (!plugins.hasPlugin('java')) {
+                def message =
+                        "j2objc plugin didn't find the 'java' plugin (which was expected).\n"+
+                        "You must use the 'java' plugin (or or one of its derivatives).  See:\n" +
+                        "https://docs.gradle.org/current/userguide/java_plugin.html\n" +
+                        "Be sure to apply the j2objc plugin after the java plugin."
```
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: Change the description as follows:

                        "j2objc plugin didn't find the 'java' plugin in the '${project.name}' project.\n"+
                        "This is a requirement for using j2objc. If you are migrating an existing\n" +
                        "Android app, it is suggested that you create a separate 'shared' project\n" +
                        "without android dependencies and gradually migrate shared code there.\n" +
                        "Within that project, first apply the 'java' then 'j2objc' plugins by adding\n" +
                        "the following lines to the build.gradle file:\n" +
                        "\n" +
                        "    apply plugin: 'java'\n" +
                        "    apply plugin: 'j2objc'\n" +
                        "\n" +
                        "    j2objcConfig {\n" +
                        "        // j2objc settings here\n" +
                        "    }\n" +
                        "\n" +
                        "More Info: https://github.com/j2objc-contrib/j2objc-gradle/#usage"

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy [hunk 3]
```diff
@@ -94,12 +98,9 @@ class J2objcPlugin implements Plugin<Project> {
...
                 description "Translates all the java source files in to Objective-C using j2objc"
                 // TODO: Once this is a proper plugin, we can switch this to use SourceSets.
-                srcFiles = files(
-                        fileTree(dir: projectDir,
-                                include: "**/*.java",
-                                exclude: relativePath(buildDir)) +
+                additionalSrcFiles = files(
                         fileTree(dir: "build/source/apt",
-                                include: "**/*.java")
+                                 include: "**/*.java")
```
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: Delete erroneously added " " before "include"

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy [hunk 4]
```diff
@@ -94,12 +98,9 @@ class J2objcPlugin implements Plugin<Project> {
                     dependsOn: 'j2objcCycleFinder') {
                 description "Translates all the java source files in to Objective-C using j2objc"
                 // TODO: Once this is a proper plugin, we can switch this to use SourceSets.
-                srcFiles = files(
-                        fileTree(dir: projectDir,
-                                include: "**/*.java",
-                                exclude: relativePath(buildDir)) +
+                additionalSrcFiles = files(
```
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: Shouldn't additionalSrcFiles also be added to the cycleFinderTask sourceSet?

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy [hunk 5]
```diff
@@ -112,10 +113,7 @@ class J2objcPlugin implements Plugin<Project> {
             tasks.create(name: "j2objcTest", type: J2objcTestTask,
                     dependsOn: 'debugTestJ2objcExecutable') {
                 description 'Runs all tests in the generated Objective-C code'
-                srcFile = file("${buildDir}/binaries/testJ2objcExecutable/debug/testJ2objc")
-                // Doesn't use 'buildDir' as missing full path with --no-package-directories flag
-                // TODO: Once this is a proper plugin, we can switch this to use SourceSets.
-                srcFiles = files(fileTree(dir: projectDir, includes: ["**/*Test.java"]))
+                testBinaryFile = file("${buildDir}/binaries/testJ2objcExecutable/debug/testJ2objc")
             }
             // Check task is added by lifecycle plugin.
```
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: Reword comment:

    // 'check' task is added by 'java' plugin, it depends on 'test' and all the other verification tasks

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy (hunk 1 above)
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: I think of the j2objc plugin as requiring a clean test / check pass of the java plugin first. Since 'j2objcTest' depends on 'cycleFinderTask', the transitive dependency means that you can remove line 119:

    lateDependsOn(project, 'j2objcTest', 'check')

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcCycleFinderTask.groovy [hunk 6]
```diff
@@ -24,15 +23,42 @@ import org.gradle.api.tasks.InputFiles
...
  */
 class J2objcCycleFinderTask extends DefaultTask {
 
-    // Java source files
     @InputFiles
-    FileCollection srcFiles
+    FileCollection getSrcFiles() {
+        // Note that translatePattern does not need
+        // to be @Inputs because they are solely inputs to this method, which
+        // is already an input.
```
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcCycleFinderTask.groovy]: Reformat comment to 3 lines instead of 2. Also change @Inputs to @Input

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcCycleFinderTask.groovy [hunk 7]
```diff
@@ -24,15 +23,42 @@ import org.gradle.api.tasks.InputFiles
...
+        def allFiles = J2objcUtils.srcDirs(project, 'main', 'java')
+        allFiles = allFiles.plus(J2objcUtils.srcDirs(project, 'test', 'java'))
+        if (project.j2objcConfig.translatePattern != null) {
+            allFiles = allFiles.matching(project.j2objcConfig.translatePattern)
```
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcCycleFinderTask.groovy]: Perfect. This is great.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcCycleFinderTask.groovy [hunk 8]
```diff
@@ -24,15 +23,42 @@ import org.gradle.api.tasks.InputFiles
...
+        }
+        return allFiles
+    }
+
+    // All input files that could affect translation output, except those in j2objc itself.
+    @InputFiles
+    FileCollection getAllInputFiles() {
+        FileCollection allFiles = srcFiles
+        if (translateSourcepaths) {
```
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcCycleFinderTask.groovy]: I prefer to make the j2objcConfig dependency more explicit by generally writing it as:

    project.j2objcConfig.translateSourcepaths

This would be consistent with the previous getSrcFiles() method. When you see "translateSourcepaths" in isolation, it's not as obvious to connect that to getTranslateSourcepaths() below. I understand what you're saying by having a more groovy style.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcCycleFinderTask.groovy [hunk 9]
```diff
@@ -24,15 +23,42 @@ import org.gradle.api.tasks.InputFiles
...
+            def translateSourcepathPaths = translateSourcepaths.split(':') as List<String>
```
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcCycleFinderTask.groovy]: I was also wondering if we should change translateSourcepathPaths from a String to String[]. Your thoughts?

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcCycleFinderTask.groovy (hunk 8 above)
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcCycleFinderTask.groovy]: Or perhaps use getTranslateSourcepaths(), to be consistent with the code elsewhere in the task.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTranslateTask.groovy [hunk 10]
```diff
@@ -0,0 +1,271 @@
+/*
```
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTranslateTask.groovy]: Looks like the file changed so much that Git didn't recognize it as a move. The SO advice it to do separate commits, firstly the move and then the edit. Just take care not to squash the commit, otherwise Git might get confused again!

http://stackoverflow.com/questions/433111/how-to-make-git-mark-a-deleted-and-a-new-file-as-a-file-move/433114#433114

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcUtils.groovy [hunk 11]
```diff
@@ -32,51 +32,22 @@ class J2objcUtils {
...
-                // For standard fileTypes 'java' and 'resources,' per contract
-                // this cannot be null.
-                return sourceSet.getProperty(fileType).srcDirs
-            }
-        }
-        return DEFAULT_SOURCE_DIRECTORIES[sourceSetName][fileType].collect({ proj.file(it) })
+        SourceSet sourceSet = proj.sourceSets.findByName(sourceSetName)
+        // For standard fileTypes 'java' and 'resources,' per contract
+        // this cannot be null.
+        return sourceSet.getProperty(fileType)
```
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcUtils.groovy]: Wonderful to get this cleaned up.

**Comment by brunobowden** [comment]
Hi @advayDev1,
Thanks for all these changes. It's wonderful to get so much of this cleaned up. I've done my initial round of comment which I think should be pretty comprehensive.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy (hunk 1 above)
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: so L119 means check depends on j2objcTest, while L90 is j2objcCycleFinder depends on test.

check is a lifecycle plugin task and so covers /all/ checks for a project.  test is for java plugin tests only, i think this overall makes sense.  namely,  all objc work (starting with j2objcCycleFinder) depends on the java plugin being fully tested (i.e. 'test'); the project being fully checked (i.e. 'check') depends on the objc work being fully tested (i.e. 'j2objcTest').

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy (hunk 2 above)
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: interesting that github generates arbitrary user-specified anchors without prefixes and such...

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy (hunk 3 above)
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: i thought this was our new alignment strategy

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy (hunk 4 above)
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: i want to keep this CL a perfect refactor in that it is functionally identical to the old way of doing things.  for some reason only translate had build/source/apt, and i'll keep it that way.

then someone who uses cycleFinder and/or build/source/apt (i use neither) can make the appropriate fix in a distinct cl.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy (hunk 5 above)
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: added, slightly modified

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcCycleFinderTask.groovy (hunk 6 above)
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcCycleFinderTask.groovy]: done

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcCycleFinderTask.groovy (hunk 8 above)
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcCycleFinderTask.groovy]: went with get, i explicitly want the return value of the @Input method.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcCycleFinderTask.groovy (hunk 9 above)
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcCycleFinderTask.groovy]: let's do it in a separate cl

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTranslateTask.groovy (hunk 10 above)
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTranslateTask.groovy]: actually this is just because #109 isn't submitted, once that is merged and i rebase it should be fine

**Comment by advayDev1** [comment]
Thanks @brunobowden - some changes completed, some changes replied to.
If you merge in #109, I'll rebase/squash this one to make changes clearer.

**Comment by advayDev1** [comment]
oh cool, you did merge #109 - one second while i do the rebase/squash...

**Comment by advayDev1** [comment]
great, ptal

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy (hunk 1 above)
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: Got it. That now makes sense to me.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy [hunk 12]
```diff
@@ -112,12 +123,10 @@ class J2objcPlugin implements Plugin<Project> {
...
-                srcFile = file("${buildDir}/binaries/testJ2objcExecutable/debug/testJ2objc")
-                // Doesn't use 'buildDir' as missing full path with --no-package-directories flag
-                // TODO: Once this is a proper plugin, we can switch this to use SourceSets.
-                srcFiles = files(fileTree(dir: projectDir, includes: ["**/*Test.java"]))
+                testBinaryFile = file("${buildDir}/binaries/testJ2objcExecutable/debug/testJ2objc")
             }
-            // Check task is added by lifecycle plugin.
+            // 'check' task is added by 'java' plugin, it depends on 'test' and all the other verification tasks,
+            // now including 'j2objcTest'.
             lateDependsOn(project, 'j2objcTest', 'check')
```
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: I think it would be best to reverse the parameters of lateDependsOn, so that before comes before after. This would match the parameter ordering of tasks.create(...). For example:

    lateDependsOn(project, 'check', 'j2objcTest')

Can be mentally translated as: **'check' dependsOn 'j2objcTest'**

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPluginExtension.groovy [hunk 13]
```diff
@@ -97,44 +78,26 @@ class J2objcPluginExtension {
...
-    // Excludes a fully qualified class name from translation.
-    // Example: translateExcludeClass 'java.lang.String'
-    def translateExcludeClass(def className) {
-        translateExcludeRegex = regexOr(translateExcludeRegex, classNameToFileRegex(className))
-    }
-    // Includes a fully qualified class name in translation.
-    // Example: translateIncludeClass 'java.lang.String'
-    def translateIncludeClass(def className) {
-        translateIncludeRegex = regexOr(translateIncludeRegex, classNameToFileRegex(className))
+        def code = cl.rehydrate(translatePattern, cl.owner, cl.thisObject)
```
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPluginExtension.groovy]: Add comment explaining what rehydrate is doing

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPluginExtension.groovy [hunk 14]
```diff
@@ -147,10 +110,27 @@ class J2objcPluginExtension {
...
+    // If no pattern is specified, all files within the sourceSet are tested.
+    PatternSet testPattern = null
+    // DSL method to conveniently configure the testPattern.
+    def testPattern(Closure cl) {
+        if (testPattern == null) {
+            testPattern = new PatternSet()
+        }
+        def code = cl.rehydrate(testPattern, cl.owner, cl.thisObject)
+        code.call()
+        return testPattern
```
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPluginExtension.groovy]: Extract methods for common code in 'testPattern()' and 'translatePattern()'

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcUtils.groovy [hunk 15]
```diff
@@ -32,51 +32,22 @@ class J2objcUtils {
...
 
-    // Retrieves the configured source directories per specification indicated
-    // in documentation of 'custom.*SrcDirs'.
-    private final static def DEFAULT_SOURCE_DIRECTORIES =
-            [main: [java: ['src/main/java'], resources: ['src/main/resources']],
-             test: [java: ['src/test/java'], resources: ['src/test/resources']]]
-
-    static def srcDirs(Project proj, String sourceSetName, String fileType) {
+    // Retrieves the configured source directories from the Java plugin SourceSets.
+    static SourceDirectorySet srcDirs(Project proj, String sourceSetName, String fileType) {
```
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcUtils.groovy]: I wonder if the method is so simple now if we shouldn't just inline it. Since the parameters are hardcoded rather than j2objcConfig, the error checking is less likely to be an issue:

    proj.sourceSets.findByName(sourceSetName).getProperty(fileType)

**Comment by brunobowden** [comment]
@advayDev1 - please respond to this last round of comments and then I'll merge it in

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy (hunk 12 above)
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPlugin.groovy]: done

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPluginExtension.groovy (hunk 13 above)
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPluginExtension.groovy]: see below

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPluginExtension.groovy (hunk 14 above)
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcPluginExtension.groovy]: found built-in util

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcUtils.groovy (hunk 15 above)
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcUtils.groovy]: fixed in a different way

**Comment by advayDev1** [comment]
@brunobowden  ptal

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcUtils.groovy (hunk 15 above)
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcUtils.groovy]: Very good point @advayDev1... it's hard to disagree when you make a good argument like that ;-)
